src/modules/brightness.rs

```diff
@@ -0,0 +1,307 @@
+        let tx = context.tx.clone();
+        let client = context.try_client::<brightness::Client>()?;
+        let icons = self.icons.clone();
+        let datasource = self.datasource.clone();
+
+        {
+            let scroll_controller =
+                EventControllerScroll::new(EventControllerScrollFlags::VERTICAL);
+
+            scroll_controller.connect_scroll(move |_, _, dy| {
+                let tx = tx.clone();
+                let client = client.clone();
+                let icons = icons.clone();
+                let datasource = datasource.clone();
+
+                spawn(async move {
+                    if let Ok((_, cur, max)) = Self::read_percentage(&client, &datasource).await {
+                        let step = ((max as f64 / 100.0) as i32).max(1); // at least modify by 1 if max < 100
+                        let new_cur = if dy > 0.0 {
+                            cur - step
+                        } else if dy < 0.0 {
+                            cur + step
+                        } else {
+                            return;
+                        };
+
+                        let new_cur = new_cur.max(0).min(max); // not using .clamp to avoid panic in case max is ever < 0
+                        let percent: f64 = cur as f64 / (max as f64) * 100.0;
+
+                        tx.send_update(BrightnessProperties {
+                            icon_name: icons.brightness_icon(percent).to_string(),
+                            screen_brightness: percent,
+                        })
+                        .await;
+
+                        match &datasource {
+                            BrightnessDataSource::Keyboard => {
+                                if let Err(e) = client.keyboard().set_brightness(new_cur).await {
+                                    tracing::error!(?e, "Could not change brightness");
+                                }
+                            }
+                            BrightnessDataSource::Login1Fs { subsystem, name } => {
+                                if let Err(e) = client
+                                    .screen_writer()
+                                    .set_brightness(
+                                        subsystem.to_string(),
+                                        name.to_string(),
+                                        new_cur as u32,
+                                    )
+                                    .await
+                                {
+                                    tracing::error!(?e, "Could not change brightness");
+                                }
+                            }
+                        };
+                    }
+                });
+
+                glib::Propagation::Proceed
+            });
+
+            button.add_controller(scroll_controller);
+        }
+
+        let rx = context.subscribe();
+
+        rx.recv_glib(
+            (&self.icons, &self.format),
+            move |(_, format), properties| {
```

Comment:
Icons is passed in as a dependency here but not used

```suggestion
            &self.format,
            move |format, properties| {
```